### ISSUE TITLE
Link not created and the Doxygen comment is placed in the output improperly

### ISSUE THREAD

**Issue by N0NB** [doxygen/doxygen#8563]
Our project https://github.com/Hamlib/Hamlib has long used Doxygen.  Earlier this year I took on the task of updating our Doxygen configuration and source file comments.  I have tried to post to the doxygen-user mailing list, but never receive a confirmation, so I am posting this here.

**Describe the bug**
As I was documenting some functions I noticed that I had some spurious text appearing. The text rendered in HTML is:

< '' – None

**Expected behavior**

That text not appear!

**Screenshots**
This online copy of the manual from the latest release shows the [output in the rig_get_mode() function](http://hamlib.sourceforge.net/manuals/4.2/group__rig.html#ga1281f19b50f2401fbfd69e7ddc561aa0).

Here is the [source code comment](https://github.com/Hamlib/Hamlib/blob/86d2b4ee6c07bb7dd2af70a51567f2bc86adf4e6/src/rig.c#L2207) to the above output.  Note, that this is not the section of code I've been working with, just the most egregious example.

I found the [offending text in one of our main header files](https://github.com/Hamlib/Hamlib/blob/86d2b4ee6c07bb7dd2af70a51567f2bc86adf4e6/include/hamlib/rig.h#L1163):

```
#define    RIG_MODE_NONE      0                         /*!< '' -- None */
```

Which immediately precedes a series of defines of bit shifted values such as:

```
#define    RIG_MODE_AM        CONSTANT_64BIT_FLAG (0)   /*!< \c AM -- Amplitude Modulation */
#define    RIG_MODE_CW        CONSTANT_64BIT_FLAG (1)   /*!< \c CW -- CW "normal" sideband */
#define    RIG_MODE_USB       CONSTANT_64BIT_FLAG (2)   /*!< \c USB -- Upper Side Band */
```

**To Reproduce**
I will see if I can work up an example in the coming days.  This project is large and complex and something elsewhere could be working in concert to cause this.

**Version**
1.9.1 packaged in the forthcoming Debian 11 release.

**Additional context**
I will assist in all ways possible.

**Comment by albert-github** [comment]
>  I have tried to post to the doxygen-user mailing list, but never receive a confirmation, so I am posting this here.

wise decision

You already mentioned a few things:
- offending define
- the part where the problem occurs

this looks all OK.

- what happens when you use (as a workaround):  /*!< \c '' -- None */
- can you post already the Doxyfile you used?

**Comment by albert-github** [comment]
In `rig.c` I see on line 2207, `/*` instead of `/**`.

**Comment by N0NB** [comment]
> In `rig.c` I see on line 2207, `/*` instead of `/**`.

Oops!  I missed that.  I see that was broken a couple of months ago.  Thanks for the catch.

I'll test and reply to your earlier message this evening.

**Comment by N0NB** [comment]
> * what happens when you use (as a workaround):  /*!< \c '' -- None */

See the attached screenshots, it appears the only change is the single quotes appear in a fixed width font:

![image](https://user-images.githubusercontent.com/555821/119209344-b708b400-ba6b-11eb-958e-9e1958cdd272.png)

and after:

![image](https://user-images.githubusercontent.com/555821/119209365-ca1b8400-ba6b-11eb-9406-4280829aae9c.png)


This is the function you pointed out was missing the second asterisk on the opening line of the comment.

> * can you post already the Doxyfile you used?

As requested:

```
$ cat hamlib.cfg 
## hamlib.cfg.in--converted into hamlib.cfg by config.status

PROJECT_NAME    = Hamlib
PROJECT_NUMBER  = 4.3~git
PROJECT_LOGO    = ../../hamlib/doc/hamlib.png

# Output
OUTPUT_DIRECTORY= .

OUTPUT_LANGUAGE = English
GENERATE_HTML   = YES
GENERATE_LATEX  = NO
GENERATE_RTF    = NO
GENERATE_MAN    = NO    # User can set to YES for local man pages
MAN_EXTENSION	= .3

# Set GENERATE_LATEX to YES to generate the needed files for the PDF manual
# and run 'make doc'.  Then 'cd latex ; make' to generate the PDF manual.
#
# See section 1.3.1.2 in ../../hamlib/README.developer for the needed LaTEX
# packages.
USE_PDFLATEX    = YES
PDF_HYPERLINKS  = YES
EXTRA_PACKAGES  = pmboxdraw

# Input
CASE_SENSE_NAMES = YES
FULL_PATH_NAMES  = NO
INPUT            = ../../hamlib/doc/index.doxygen \
                   ../../hamlib/include/hamlib/ \
                   ../../hamlib/src/

EXCLUDE          = ../../hamlib/src/amp_conf.h \
                   ../../hamlib/include/hamlib/ampclass.h \
                   ../../hamlib/include/hamlib/rotclass.h

INCLUDE_PATH     = ../../hamlib/include

EXAMPLE_PATH     = ../../hamlib/tests/testrig.c \
                   ../../hamlib \
                   ../../hamlib/scripts \
                   ../scripts


QUIET            = YES

HTML_FOOTER           = ../../hamlib/doc/footer.html
HTML_EXTRA_STYLESHEET = ../../hamlib/doc/hamlib.css

EXTRACT_ALL           = NO
EXTRACT_STATIC        = NO
HIDE_UNDOC_MEMBERS    = NO
SHOW_INCLUDE_FILES    = YES
INHERIT_DOCS          = YES
ENABLED_SECTIONS      = ""
PREDEFINED            = DOXYGEN "DOC_HIDDEN"
JAVADOC_AUTOBRIEF     = NO
OPTIMIZE_OUTPUT_FOR_C = YES

MAN_LINKS             = NO
MACRO_EXPANSION       = YES
```

**Comment by albert-github** [comment]
I've been able to reproduce the problem and create a small example:

**rig.h**
```
#define    RIG_MODE_NONE      0                         /*!< '' -- None */
```
**rig.c**
```
/// \file

#include "rig.h"

int HAMLIB_API rig_get_mode(int *width)
{
 *mode = RIG_MODE_NONE;
}
```
**Doxyfile**
```
QUIET = YES
INPUT = rig.c rig.h
MACRO_EXPANSION=YES
```

How does doxygen see this, for this we need the results of the preprocessing as done by doxygen (`doxygen -d preprocessor`):
```
#include rig.h: parsing...
Preprocessor output of D:/speeltuin/issue_8563/ex_err/rig.c (size: 90 bytes):
---------
00001 /// \file
00002
00003
00004
00005 int HAMLIB_API rig_get_mode(int *width)
00006 {
00007  *mode =  0 /*!< '' -- None */ ;
00008 }
00009
---------
Macros accessible in this file (D:/speeltuin/issue_8563/ex_err/rig.c):
---------
RIG_MODE_NONE
---------
```
so doxygen doesn't see the `RIG_MODE_NONE` in the code anymore but the real value ** including** the comment there.
This comment is giving some problems as doxygen will add this to the last documentable item seen, in this case the function `rig_get_mode`.

Solutions:
- setting `MACRO_EXPANSION=NO`, not the best solution as in that case all defines won't be replaced
- replacing
   ```
   #define    RIG_MODE_NONE      0                         /*!< '' -- None */
   ```
   by
   ```
   /*! '' -- None */
   #define    RIG_MODE_NONE      0
   ```

I think it would be worthwhile to mention this artifact in the documentation in the paragraph: "Putting documentation after members?

Example: [example.tar.gz](https://github.com/doxygen/doxygen/files/6526203/example.tar.gz)

**Comment by N0NB** [comment]
Thanks, Albert.

This seems to work well:

```
/**
 * \def RIG_MODE_NONE
 * \c '' -- None
 */
#define    RIG_MODE_NONE      0
```

Oddly, there are other constructs like this one for other `#define` groups.  Perhaps I'll run into this issue with them.

I wonder, is it the value of `0` that is tripping up Doxygen?

**Comment by albert-github** [comment]
The other defines like `RIG_MODE_AM` will suffer from a similar problem, this can be easily tested by taking the original `rig.h` and extending the small example `rig.c` by
```
*mode = RIG_MODE_AM;
```
etc.
Most likely the other defines are not used in a documentation part but e.g. in a static function / struct.

I think there are also some other places that suffer from some problems in respect to "documentation after" I see e.g `struct { /*!< */`.

A bit of an overview can be given by running:
```
grep "&lt;" -r | grep -v "&lt; *&lt;"| grep -v "&lt;.*&gt;" | grep -v "_source"
```
in the html directory, which gives a bit of an search indication.

**Comment by N0NB** [comment]
Thanks for the tips and help Albert.  I will be aware of the possibility of trailing documentation going forward and will put a link to this issue in our developer documentation.

As I know what is likely happening and how to deal with it, this issue can be closed but I'll let you or another maintainer decide when to close this.

**Comment by albert-github** [comment]
I've just pushed a proposed documentation patch, pull request #8566

**Comment by albert-github** [comment]
A bit corrected text has been integrated in master on github (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.2.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
